File: go/ids.go
package cinchor

import (
	"crypto/sha256"
	"encoding/binary"

	omne "github.com/OmneDAO/omne-sdks/go"
)

func u64be(n uint64) []byte {
	b := make([]byte, 8)
	binary.BigEndian.PutUint64(b, n)
	return b
}

// DeriveCapabilityID = sha256(principal ‖ agent ‖ u64be(nonce) ‖ u64be(createdAt)) → om1z.
// The contract is agnostic to this derivation; it must match the TS/Python SDKs.
func DeriveCapabilityID(principal, agent string, nonce, createdAt uint64) (string, error) {
	p, err := omne.FromOmneAddress(principal)
	if err != nil {
		return "", err
	}
	a, err := omne.FromOmneAddress(agent)
	if err != nil {
		return "", err
	}
	pre := make([]byte, 0, 32+32+16)
	pre = append(pre, p...)
	pre = append(pre, a...)
	pre = append(pre, u64be(nonce)...)
	pre = append(pre, u64be(createdAt)...)
	h := sha256.Sum256(pre)
	return omne.ToOmneAddress(h[:])
}

// DeriveActionID = sha256("cinchor.action" ‖ capabilityID ‖ u64be(currentTime) ‖
// u64be(amount) ‖ u64be(seq)) → om1z. The on-chain idempotency key for
// record_action: it must be STABLE for a given logical action so a worker
// resubmit (after a crash, possibly with a reseeded nonce) dedups to
// exactly-once. The gateway derives it once (seq = a per-op nonce), freezes it
// on the operation, and replays the same value; the standalone blocking path
// derives a fresh one per call (no resubmit, so uniqueness suffices).
func DeriveActionID(capabilityID string, currentTime, amount, seq uint64) (string, error) {
	c, err := omne.FromOmneAddress(capabilityID)
	if err != nil {
		return "", err
	}
	pre := make([]byte, 0, 14+32+24)
	pre = append(pre, []byte("cinchor.action")...)
	pre = append(pre, c...)
	pre = append(pre, u64be(currentTime)...)
	pre = append(pre, u64be(amount)...)
	pre = append(pre, u64be(seq)...)
	h := sha256.Sum256(pre)
	return omne.ToOmneAddress(h[:])
}

// CounterpartyKey = sha256(capabilityID ‖ counterparty) → om1z. Capability-scoped;
// minting an allowed counterparty and enforcing an action use the same derivation.
func CounterpartyKey(capabilityID, counterparty string) (string, error) {
	c, err := omne.FromOmneAddress(capabilityID)
	if err != nil {
		return "", err
	}
	cp, err := omne.FromOmneAddress(counterparty)
	if err != nil {
		return "", err
	}
	pre := make([]byte, 0, 64)
	pre = append(pre, c...)
	pre = append(pre, cp...)
	h := sha256.Sum256(pre)
	return omne.ToOmneAddress(h[:])
}

File: py/src/cinchor/ids.py
"""Deterministic capability/counterparty id derivations (port of address.ts).

The contract is agnostic to these derivations — it only compares stored keys —
so the SDK owns them. They must match the TS/Go SDKs byte-for-byte so a
capability minted by one SDK is addressable by another.
"""

from __future__ import annotations

import hashlib

from omne_sdk.address import from_omne_address, to_omne_address


def _u64be(n: int) -> bytes:
    return int(n).to_bytes(8, "big")


def derive_capability_id(principal: str, agent: str, nonce: int, created_at: int) -> str:
    """sha256(principal ‖ agent ‖ u64be(nonce) ‖ u64be(created_at)) → om1z."""
    preimage = (
        from_omne_address(principal)
        + from_omne_address(agent)
        + _u64be(nonce)
        + _u64be(created_at)
    )
    return to_omne_address(hashlib.sha256(preimage).digest())


def counterparty_key(capability_id: str, counterparty: str) -> str:
    """Per-(capability, counterparty) allowlist key: sha256(cap ‖ counterparty) → om1z.

    Capability-scoped (the id is in the preimage) so there is no cross-capability
    collision. Minting an allowed counterparty and enforcing an action MUST use
    the same derivation.
    """
    preimage = from_omne_address(capability_id) + from_omne_address(counterparty)
    return to_omne_address(hashlib.sha256(preimage).digest())
